### ISSUE TITLE
I have tried opening ITV hub and selected a program, but although Version 1.10.2 (20180314002) allows me to add the programme clicking on the use current webpage, GIA still won't allow me to download a programme from ITV. 

### ISSUE THREAD

**Issue by Swilly6vic** [Ascoware/get-iplayer-automator#125]
**This issue tracker is not a discussion forum, so don't use it as such.  If you wish to discuss the BBC, iPlayer or anything else that is not related to an actual problem with the current release of GiA, post in an appropriate forum, not in this issue tracker.**

**STOP! DO NOT submit a ticket if you are reporting that you are unable to download a program from outside the UK**.

If you need an explanation, read [this](https://github.com/Ascoware/get-iplayer-automator/wiki/No-Support-for-GiA-Outside-the-UK). If you ignore this instruction, your ticket will be closed with no further action. You can fib about your location if you like, but you will only be wasting your time. If you are using GiA within the UK, continue with the instructions below.

Before creating a new ticket, search the current open and closed tickets to see if your issue has already been reported. Use the search box at the top of the page to do so. You can also use the links below to view open and closed tickets in list form.

#### [Closed tickets](https://github.com/Ascoware/get-iplayer-automator/issues?q=is%3Aissue+is%3Aclosed)

If you create a ticket for an issue that has already been reported, your ticket will be closed with no further action. If you are certain your issue has not yet been reported, continue with the instructions below.

Create a succinct title describing your issue in the `Title` field above, then provide the information requested below. If you do not provide the requested information, your ticket will be closed with no further action.

If you cannot download a program, provide a verbose log of the attempt. See [Verbose Log](https://github.com/Ascoware/get-iplayer-automator/wiki/Verbose-Log) in wiki for instructions on posting a verbose log for your ticket. **Your verbose log should show the results from downloading ONLY ONE program. DO NOT post a massive log covering multiple multiple program downloads. IMPORTANT: DO NOT paste your verbose log into your issue report. If you do, expect it to be deleted.**

If GiA crashes, provide the OS X crash log by copying the information from Crash Reporter into a file or locating the crash log in `$HOME/Library/Logs/DiagnosticReports`. See [Verbose Log](https://github.com/Ascoware/get-iplayer-automator/wiki/Verbose-Log) in wiki for instructions on posting a crash log for your ticket. **IMPORTANT: DO NOT paste your crash log into your issue report. If you do, expect it to be deleted.**

**Delete the instructions above before submitting your ticket.**

---------------

*What steps will reproduce the problem? Be specific, and provide as much detail as possible.*

*What is the expected output? What do you see instead?*

*What version of GiA are you using? On what version of OS X?*

*Are you using a web proxy? DNS proxy? VPN? VPS?*

*What program are you attempting to download (full title, series and episode)? **Provide only ONE program as an example.** *

*Please provide any additional information below.*

**Comment by Lordsnooty05** [comment]
Swilly6vic please read:-

"Searches aren't working right now because the cache has no program IDs or URLs.

It looks like ITV have changed the layout of their program pages, and as a result the ITV scanning is now broken. It's not possible to use the 'Use current webpage' either, as the parsing code can't find the program data.

I'll investigate ASAP.".......Scott

Please give him time to work on it, it appears the fault lies with the redesign of ITV Player. Please try and read through the issues before starting a new issue which has and is essentially being dealt with by Scott. All this does is put pressure on him to answer what is a question that others have put to him. Thanks

**Comment by dvioletta** [comment]
I might have information that can help after a conversation with ITVhub support on twitter. It seems that some programs can only be watched on Internet Explorer version 11.

**Comment by jchhenderson** [comment]
That's is correct.

Another example (I wrote it in #115) is that "Lethal Weapon" is only available on their website via Internet Explorer or Microsoft Edge, but due to technical reasons it is not available via any other browsers.

...I was wondering whether GiA could emulate IE in order to access the content.

**Comment by jefro108** [comment]
I recently found MacYTDL https://forum.videohelp.com/threads/384566-youtube-dl-GUI-for-Apple-Macs#post2513526 to download the ITV programmes that don't work in Get iPlayer Automator - although it doesn't work for Lethal Weapon

**Comment by dvioletta** [comment]
That is a pity it is pretty much only one episode of Lethal Weapon that I am missing from my current run. Episode 7 from season 1

**Comment by skovatch** [comment]
Fixed in 1.11, though Lethal Weapon still won't work. See #126 for more information.

**Comment by Vangelis66** [comment]
... Some additional info with regards to "Lethal Weapon": 

@dvioletta wrote: 
> It seems that some programs can only be watched on Internet Explorer version 11.

@jchhenderson wrote: 
> "Lethal Weapon" is only available on their website via Internet Explorer (11) or Microsoft Edge, 
> but due to technical reasons it is not available via any other browsers.
> (snip)
> I was wondering whether GiA could emulate IE in order to access the content.

@skovatch wrote: 
> though Lethal Weapon still won't work

... As I've already mentioned several other times in this tracker, **all** the VOD streams ITVHub currently serves to **desktop** browsers are encrypted (**DRM**) and are of no use to GIA... 
The streams are typically of the MPEG-DASH type (with Common Encryption, cenc), various browsers on various platforms have differing support for decoding/decrypting those streams (e.g. you can have a quick look at [this](https://developer.bitmovin.com/hc/en-us/articles/115001445814-Browser-DRM-Support) article). 

 It would appear that the "**Lethal Weapon**" series (and other programmes like that) has been licenced as an on-demand stream (on desktop browsers) only when the **PlayReady** DRM encryption is being used; PlayReady is a Microsoft owned DRM system, applicable in MS's IE11 (Win8.1+) / Edge (Win10). 

On most other desktop browsers (Chromium + Mozilla families), typical ITVHub VOD uses Google owned Widevine+ClearKey DRM decryption for DASH streams; like Scott, I too conducted tests with my FirefoxESR 52.7.2 (that won't play LW) and was successful in making ITVHub think I have IE11 (by using a global useragent override in Firefox); although Firefox's Web console would display DASH video+audio fragments being downloaded in the background, these couldn't be decrypted for lack of PlayReady support in Firefox...

On desktop Safari, on iOS and possibly other mobile platforms, ITVHub serves VOD streams with AES-128 encryption; this type of encryption has been reversed-engineered (cracked), hence these streams are "good" for youtube-dl/FFmpeg/GiA etc. The problem with LW is that there are no such streams available for it... Only stream source for LW is the old RTMP infrastructure for ITVHub (_yet another reason not to remove it from GiA_) but, as we know, not all episodes have RTMP streams prepared for them...
E.g. on latest episode of Series 2 (S02E10), yt-dl reports: 
```
youtube-dl -F "https://www.itv.com/hub/lethal-weapon/2a4808a0028"
[ITV] 2a4808a0028: Downloading webpage
[ITV] 2a4808a0028: Downloading XML
[ITV] 2a4808a0028: Downloading JSON metadata
WARNING: Unable to download JSON metadata: HTTP Error 404: Not Found
ERROR: No video formats found; (redacted)
```
**No HLS**, **No RTMP** streams; OTOH, for S02E09: 
```
youtube-dl -F "https://www.itv.com/hub/lethal-weapon/2a4808a0027"
[ITV] 2a4808a0027: Downloading webpage
[ITV] 2a4808a0027: Downloading XML
[ITV] 2a4808a0027: Downloading JSON metadata
WARNING: Unable to download JSON metadata: HTTP Error 404: Not Found
[info] Available formats for 2a4808a0027:
format code  extension  resolution note
rtmp-600     flv        unknown     600k
rtmp-800     flv        unknown     800k
rtmp-1200    flv        unknown    1200k
rtmp-1500    flv        unknown    1500k
rtmp-1800    flv        unknown    1800k  (best)
```
that is **No HLS** but **Yes RTMP** streams!

@jefro108 wrote: 
> though Lethal Weapon **still** won't work

 Since MacYTDL is but a frontend to the Python script that is youtube-dl, MacYTDL can go as far as yt-dl can...
